### PR TITLE
Bug 1756453: Separate upgrade flags for safety instead of abusing force

### PR DESCRIPTION
**Pull request by smarterclayton**
The --force flag is dangerous and potentially allows untrusted content to be upgraded to accidentally. Instead, introduce two new flags `--allow-explicit-upgrade` (for upgrading to something not in availableVersions) and `--allow-upgrade-with-warnings` (for upgrading when another upgrade is in progress or the cluster is reporting an error) and remove those checks from `--force`.

While this is an API change, it is necessary to ensure that users do not accidentally get access to untrusted content when performing upgrades across major versions in advance of graph updates, or when they are upgrading in disconnected environments.

/assign @jwforres
@eparis as discussed today

This should be a candidate for backport to 4.2 and 4.1. While it changes flag behavior, it will only break things in a "safe" way. This addresses a significant security challenge as we have introduced new meanings for --force.

```
# install 4.1.15

$ oc adm upgrade --to-image quay.io/openshift-release-dev/ocp-release:4.1.16
error: The requested upgrade image is not one of the available updates, you must pass --allow-explicit-upgrade to continue
$ oc adm upgrade --to-image quay.io/openshift-release-dev/ocp-release:4.1.16 --allow-explicit-upgrade

... // wait

$ oc adm upgrade --to-image quay.io/openshift-release-dev/ocp-release:4.1.7
error: The requested upgrade image is not one of the available updates, you must pass --allow-explicit-upgrade to continue
$ oc adm upgrade --to-image quay.io/openshift-release-dev/ocp-release:4.1.7 --allow-explicit-upgrade
error: Already upgrading, pass --allow-upgrade-with-warnings to override.

  Reason:
  Message: Working towards 4.1.16: 13% complete

$ oc adm upgrade --to-image quay.io/openshift-release-dev/ocp-release:4.1.7 --allow-explicit-upgrade --allow-upgrade-with-warnings
Updating to release image quay.io/openshift-release-dev/ocp-release:4.1.7
```

### CODE REVIEW

**Comment by smarterclayton** [comment]
@abhinavdahiya you and I have discussed this, as part of ensuring 4.1 to 4.2 upgrades go smoothly we want to allow people in 4.1 to upgrade to 4.2 before we publish the edge *without* using --force to bypass content checks.

File: pkg/cli/admin/upgrade/upgrade.go [hunk 1]
```diff
@@ -70,6 +78,8 @@ func New(f kcmdutil.Factory, parentName string, streams genericclioptions.IOStre
 	flags.BoolVar(&o.ToLatestAvailable, "to-latest", o.ToLatestAvailable, "Use the next available version")
 	flags.BoolVar(&o.Clear, "clear", o.Clear, "If an upgrade has been requested but not yet downloaded, cancel the update. This has no effect once the update has started.")
 	flags.BoolVar(&o.Force, "force", o.Force, "Upgrade even if an upgrade is in process or other error is blocking update.")
+	flags.BoolVar(&o.AllowExplicitUpgrade, "allow-explicit-upgrade", o.AllowExplicitUpgrade, "Upgrade even if the upgrade target is not listed in the available versions list.")
+	flags.BoolVar(&o.AllowUnsafeUpgrade, "allow-unsafe-upgrade", o.AllowUnsafeUpgrade, "Upgrade even if an upgrade is in process or the cluster is error is blocking update.")
```
Comment (jwforres) [inline pkg/cli/admin/upgrade/upgrade.go]: discussed changing this one to allow-upgrade-when-unhealthy

Comment (smarterclayton) [reply pkg/cli/admin/upgrade/upgrade.go]: Trying `--allow-upgrade-with-warnings`?

**Comment by abhinavdahiya** [comment]
I like this.

LGTM

File: pkg/cli/admin/upgrade/upgrade.go [hunk 2]
```diff
@@ -70,6 +80,8 @@ func New(f kcmdutil.Factory, parentName string, streams genericclioptions.IOStre
 	flags.BoolVar(&o.ToLatestAvailable, "to-latest", o.ToLatestAvailable, "Use the next available version")
 	flags.BoolVar(&o.Clear, "clear", o.Clear, "If an upgrade has been requested but not yet downloaded, cancel the update. This has no effect once the update has started.")
 	flags.BoolVar(&o.Force, "force", o.Force, "Upgrade even if an upgrade is in process or other error is blocking update.")
+	flags.BoolVar(&o.AllowExplicitUpgrade, "allow-explicit-upgrade", o.AllowExplicitUpgrade, "Upgrade even if the upgrade target is not listed in the available versions list.")
+	flags.BoolVar(&o.AllowUpgradeWithWarnings, "allow-upgrade-with-warnings", o.AllowUpgradeWithWarnings, "Upgrade even if an upgrade is in process or the cluster is error is blocking update.")
```
Comment (crawford) [inline pkg/cli/admin/upgrade/upgrade.go]: "is error is blocking"?

Comment (crawford) [inline pkg/cli/admin/upgrade/upgrade.go]: Other potential colors for the shed:

`--allow-unsound-upgrade`
`--allow-impatient-upgrade`

**Comment by abhinavdahiya** [comment]
There also this bug https://bugzilla.redhat.com/show_bug.cgi?id=1713263 that states that `--to-latest` and `--force` don't work together correctly? Is that expected @smarterclayton ?

**Comment by smarterclayton** [comment]
That is not expected, are you going to track that separately or want me to throw it in here?  I don't have an issue doing that.

**Comment by abhinavdahiya** [comment]
> That is not expected, are you going to track that separately or want me to throw it in here? I don't have an issue doing that.

just wanted to get the expectation esp if it covers the new flags too..., we are already tracking this and we will work to fix it separately.

File: pkg/cli/admin/upgrade/upgrade.go (hunk 2 above)
Comment (smarterclayton) [reply pkg/cli/admin/upgrade/upgrade.go]: Fixed, thanks

**Pull request merged**